### PR TITLE
chore: extend types

### PR DESCRIPTION
**Pull request by juliusmarminge**
Another follow up on #659

### CODE REVIEW

File: packages/uploadthing/src/internal/handler.ts [hunk 1]
```diff
@@ -309,10 +309,11 @@ export const buildRequestHandler = <
             (f) =>
               isObject(f) &&
               typeof f.name === "string" &&
-              typeof f.size === "number",
+              typeof f.size === "number" &&
+              typeof f.type === "string",
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:313]: idk if we should send this from client or use lookup using filename, but i think it should be included in the `opts.files` you get in `.middleware()`.

File: packages/uploadthing/src/types.ts [hunk 2]
```diff
@@ -87,26 +71,39 @@ export type GenerateUploaderOptions = {
...
-export type UploadThingResponse = (PSPResponse | MPUResponse)[];
+/**
+ * When the client has uploaded a file and polled for data returned by `.onUploadComplete()`
+ */
+export interface ClientUploadedFileData<T> extends UploadedFileData {
+  /**
+   * Matches what's returned from the serverside `onUploadComplete` callback
+   */
+  serverData: T;
+}
```
Comment (juliusmarminge) [inline packages/uploadthing/src/types.ts:109]: this is the main change i wanted to make

We start with what the client sends (FileUploadData), then once middleware has run, we may have customId (FileUploadDataWithCustomId), then when the file has uplaoded we get `key` and `url` from UT (UploadedFileData). Lastly, the client polls for server data (ClientUploadedFileData)

File: packages/uploadthing/src/client.ts [hunk 3]
```diff
@@ -144,15 +116,16 @@ const uploadFilesInternal = async <
...
       name: file.name,
       size: file.size,
+      type: file.type,
       key: presigned.key,
-
-      serverData,
       url: "https://utfs.io/f/" + presigned.key,
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      serverData: serverData as any,
+      customId: null,
```
Comment (juliusmarminge) [inline packages/uploadthing/src/client.ts:124]: should this be filled in?

File: packages/uploadthing/src/internal/dev-hook.ts [hunk 4]
```diff
@@ -28,7 +28,16 @@ export const conditionalDevServer = async (opts: {
...
+    async (json: {
+      fileData: {
+        callbackUrl: string;
+        callbackSlug: string;
+        metadata: string | null;
+        fileName: string;
+        fileSize: number;
+        customId: string | null;
+      };
+    }) => {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/dev-hook.ts:40]: decided to inline this - not used anywhere else and very specific to the callback simulation.

File: packages/uploadthing/src/internal/handler.ts [hunk 5]
```diff
@@ -474,16 +475,13 @@ export const buildRequestHandler = <
 
         return {
           cleanup: promise,
-          body: parsedResponse.map((x) => ({
-            ...x,
-            pollingUrl: generateUploadThingURL(`/api/serverCallback`),
-          })),
+          body: parsedResponse satisfies UTEvents["upload"]["out"],
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:478]: we handle this mapping on infra, no need for mapping
CleanShot 2024-03-03 at 19 53 08@2x

**Review by markflorkowski — APPROVED** [review]

**Pull request merged**